frame/support/procedural/src/pallet/expand/hooks.rs

```diff
@@ -82,6 +82,36 @@ pub fn expand_hooks(def: &mut Def) -> proc_macro2::TokenStream {
 		proc_macro2::TokenStream::new()
 	};
 
+	// If a storage version is set, we should ensure that the storage version on chain matches the
+	// current storage version. This assumes that `Executive` is running custom migrations before
+	// the pallets are called.
+	let post_storage_version_check = if def.pallet_struct.storage_version.is_some() {
+		quote::quote! {
+			let on_chain_version = <Self as #frame_support::traits::GetStorageVersion>::on_chain_storage_version();
+			let current_version = <Self as #frame_support::traits::GetStorageVersion>::current_storage_version();
+
+			if on_chain_version != current_version {
```

Comment:
I see this does works for pallets without storage versions. But why?  
I thought it would try to compare `NoStorageVersionSet` which should error.

Reply:
Check line 88 to see why :P The code is only added if there is a storage version.